### PR TITLE
Add docs for distributed inference

### LINKED ISSUE #13187: distributed inference
We need to support distributed inference...

In addition, for now, let's make sure to add a warning when a user runs predict with accelerators > 1:

```bash
“warning: you’re running .predict with devices > 1. predictions across multiple {self.accelerator}s is not yet supported.”

# prints like
“warning: you’re running .predict with devices > 1. predictions across multiple GPUs is not yet supported.”
```

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Related #13187

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @Felonious-Spellfire

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

File: docs/source-pytorch/deploy/production_basic.rst [hunk 1]
```diff
@@ -71,10 +71,31 @@ When you need to add complicated pre-processing or post-processing logic to your
...
+    from pytorch_lightning.callbacks import BasePredictionWriter
+
+
+    class CustomWriter(BasePredictionWriter):
+        def __init__(self, output_dir, write_interval):
+            super().__init__(write_interval)
+            self.output_dir = output_dir
+
+        def write_on_epoch_end(self, trainer, pl_module, predictions, batch_indices):
+            # this will create N files in `output_dir` each containing the predictions of it's respective rank
```
Comment (carmocca) [inline docs/source-pytorch/deploy/production_basic.rst:88]: This mentions N, but not what N is. I suggest replacing it for number of processes.

File: docs/source-pytorch/deploy/production_basic.rst [hunk 2]
```diff
@@ -71,10 +71,31 @@ When you need to add complicated pre-processing or post-processing logic to your
...
+            torch.save(predictions, os.path.join(self.output_dir, f"predictions_{trainer.global_rank}.pt"))
+
+            # optionally, you can also save `batch_indices` to get the information about the data index
+            # from your prediction data
+            torch.save(predictions, os.path.join(self.output_dir, f"batch_indices_{trainer.global_rank}.pt"))
```
Comment (carmocca) [inline docs/source-pytorch/deploy/production_basic.rst:93]:
```suggestion
            torch.save(batch_indices, os.path.join(self.output_dir, f"batch_indices_{trainer.global_rank}.pt"))
```

File: src/pytorch_lightning/callbacks/prediction_writer.py [hunk 3]
```diff
@@ -52,23 +54,55 @@ class BasePredictionWriter(Callback):
...
+                super().__init__(write_interval)
+                self.output_dir = output_dir
+
+            def write_on_epoch_end(self, trainer, pl_module, predictions, batch_indices):
+                # this will create N files in `output_dir` each containing the predictions of it's respective rank
+                torch.save(predictions, os.path.join(self.output_dir, f"predictions_{trainer.global_rank}.pt"))
+
+                # optionally, you can also save `batch_indices` to get the information about the data index
+                # from your prediction data
+                torch.save(predictions, os.path.join(self.output_dir, f"batch_indices_{trainer.global_rank}.pt"))
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/prediction_writer.py:94]: Reminder to apply the same changes as above.

**Review by carmocca — COMMENTED** [review]
This PR shouldn't close the linked issue. As it's for adding support.

Also, I would suggest that this PR also adds the warning as suggested in the linked issue

**Comment by rohitgr7** [comment]
> This PR shouldn't close the linked issue. As it's for adding support.

> Also, I would suggest that this PR also adds the warning as suggested in the linked issue

maybe I am missing something, but isn't this distributed inference?

**Comment by carmocca** [comment]
William's request was that we add automagic support for returning the predictions from all processes in both ddp and ddp spawn, which will have some caveats. Without relying on a Callback to write them to disk. Doing so wouldn't necessarily replace the callback.

**Review by awaelchli — APPROVED** [review]
Great @rohitgr7 

I agree this shouldn't close the linked issue. We have briefly discussed some directions to go in, we need to write it up and post more info on the issue. To invest serious time in this, we need to put it on the roadmap first though.

File: docs/source-pytorch/deploy/production_basic.rst (hunk 1 above)
Comment (rohitgr7) [reply docs/source-pytorch/deploy/production_basic.rst]: done!

**Comment by donglihe-hub** [comment]
Hi Rohit,

I appreciate your effort in making distributed inference possible. However, I'm wondering if it is now possible to simply gather predictions from all processes without storing them first (as these predictions are temporary results)?

**Pull request merged**